flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java

```diff
@@ -886,13 +886,15 @@ public void clearJobListener() {
 			"Cannot find compatible factory for specified execution.target (=%s)",
 			configuration.get(DeploymentOptions.TARGET));
 
-		return executorFactory
+		CompletableFuture<JobClient> jobClient = executorFactory
 			.getExecutor(configuration)
-			.execute(plan, configuration)
-			.thenApply(jobClient -> {
-				jobListeners.forEach(jobListener -> jobListener.onJobSubmitted(jobClient));
-				return jobClient;
-			});
+			.execute(plan, configuration);
+
+		jobClient.thenAcceptAsync(innerJobClient -> {
```

Comment:
We should call the `Async` version if we want to asynchronously return `onJobSubmitted`. It's not only we returns in the same thread in `JobClusterExecutor` implementation but `CompletableFuture.thenAccept` is always possibly called in current thread if the future is completed when `thenXXX` called.